### PR TITLE
AcceptToMemoryPoolWorker() and relative functions

### PR DESCRIPTION
**Pull request by qshuai**
@jjz @ludete PTAL

### CODE REVIEW

File: blockchain/Validation.go [hunk 1]
```diff
@@ -2248,19 +2325,476 @@ var (
...
+		}
+
+		// Do all inputs exist?
+		for _, txin := range ptx.Ins {
+			if !gcoinsTip.HaveCoinInCache(txin.PreviousOutPoint) {
+				coinsToUncache = append(coinsToUncache, txin.PreviousOutPoint)
+			}
+
+			if !cache.HaveCoin(txin.PreviousOutPoint) {
+				if *missingInputs {
```
Comment (ludete) [inline blockchain/Validation.go]: error

File: blockchain/Validation.go [hunk 2]
```diff
@@ -2248,19 +2325,476 @@ var (
...
+	// MAX_BLOCK_SIGOPS_PER_MB; we still consider this an invalid rather
+	// than merely non-standard transaction.
+	if uint(sigOpsCount) > policy.MAX_STANDARD_TX_SIGOPS {
+		ret = state.Dos(0, false, model.REJECT_NONSTANDARD, "bad-txns-too-many-sigops",
+			false, strconv.Itoa(sigOpsCount))
+		return
+	}
+
+	var defaultRelaypriority int64
+	if DEFAULT_RELAYPRIORITY {
```
Comment (ludete) [inline blockchain/Validation.go]: convert error

File: blockchain/Validation.go [hunk 3]
```diff
@@ -2248,19 +2325,476 @@ var (
...
+			false, "")
+		return
+	}
+
+	// Continuously rate-limit free (really, very-low-fee) transactions.
+	// This mitigates 'penny-flooding' -- sending thousands of free
+	// transactions just to be annoying or make others' transactions take
+	// longer to confirm.
+	if limitFree && modifiedFees < minFeeRate {
+		var freeCount float64
```
Comment (ludete) [inline blockchain/Validation.go]: static variable life is whole program

File: blockchain/Validation.go [hunk 4]
```diff
@@ -2248,19 +2325,476 @@ var (
...
+
+	if err := pool.CalculateMemPoolAncestors(entry, setAncestors, uint64(limitAncestors), uint64(limitAncestorSize),
+		uint64(limitDescendants), uint64(limitDescendantSize), true); err != nil {
+		ret = state.Dos(0, false, model.REJECT_NONSTANDARD, "too-long-mempool-chain",
+			false, err.Error())
+		return
+	}
+
+	var scriptVerifyFlags = int64(policy.STANDARD_SCRIPT_VERIFY_FLAGS)
+	if msg.ActiveNetParams.RequireStandard {
```
Comment (ludete) [inline blockchain/Validation.go]: error

File: blockchain/Validation.go [hunk 5]
```diff
@@ -2248,19 +2325,476 @@ var (
...
+	// This is also useful in case of bugs in the standard flags that cause
+	// transactions to pass as valid when they're actually invalid. For
+	// instance the STRICTENC flag was incorrectly allowing certain CHECKSIG
+	// NOT scripts to pass, even though they were invalid.
+	//
+	// There is a similar check in CreateNewBlock() to prevent creating
+	// invalid blocks (using TestBlockValidity), however allowing such
+	// transactions into the mempool can be exploited as a DoS attack.
+	currentBlockScriptVerifyFlags := GetBlockScriptFlags(chainActive.Tip(), params) // todo confirm params
+	if !CheckInputsFromMempoolAndCache(ptx, state, &cache, pool, currentBlockScriptVerifyFlags, true, ptx) {
```
Comment (ludete) [inline blockchain/Validation.go]: the lass element pass error

File: blockchain/Validation.go [hunk 6]
```diff
@@ -2248,19 +2325,476 @@ var (
...
+func GetP2SHSigOpCount(tx *model.Tx, cache *utxo.CoinsViewCache) int {
+	if tx.IsCoinBase() {
+		return 0
+	}
+
+	sigOps := 0
+	for _, txin := range tx.Ins {
+		prevout := cache.GetOutputFor(txin)
+		if prevout.Script.IsPayToScriptHash() {
+			count, _ := prevout.Script.GetSigOpCount()
```
Comment (ludete) [inline blockchain/Validation.go]: I guess here is error

File: blockchain/Validation.go [hunk 7]
```diff
@@ -2465,3 +3002,35 @@ func GuessVerificationProgress(data *msg.ChainTxData, index *BlockIndex) float64
 
 	return float64(index.ChainTx) / txTotal
 }
+
+// c++ timedata.cpp
+var mockTime = 0
```
Comment (ludete) [inline blockchain/Validation.go]: these have write

File: blockchain/Chain.go [hunk 8]
```diff
@@ -4,6 +4,10 @@ import (
 	"sort"
 )
 
+var chainActive = Chain{
```
[the remark this answers is not in the document]
Comment (qshuai) [reply blockchain/Chain.go]: ok

File: mempool/Mempool.go [hunk 9]
```diff
@@ -1062,3 +1063,55 @@ func (mempool *Mempool) GetSortedDepthAndScore() []*TxMempoolEntry {
...
+			*coin = *utxo.NewCoin(ptx.Outs[point.Index], MEMPOOL_HEIGHT, false)
+			return true
+		}
+		return false
+	}
+
+	return m.Base.GetCoin(point, coin) && !coin.IsSpent()
+}
+func (m *CoinsViewMemPool) HaveCoin(point *model.OutPoint) bool {
+	return m.Mpool.Exists(point.Hash) || m.Base.HaveCoin(point)
```
[the remark this answers is not in the document]
Comment (qshuai) [reply mempool/Mempool.go]: ok

File: blockchain/Validation.go [hunk 10]
```diff
@@ -41,18 +43,29 @@ const (
...
+	// REJECT_HIGHFEE Too high fee. Can not be triggered by P2P transactions
+	REJECT_HIGHFEE = 0x100
+	// REJECT_ALREADY_KNOWN Transaction is already known (either in mempool or blockchain)
+	REJECT_ALREADY_KNOWN = 0x101
+	// REJECT_CONFLICT Transaction conflicts with a transaction already known
+	REJECT_CONFLICT = 0x102
+)
 
 var (
 	gsetDirtyBlockIndex *algorithm.Set
```
[the remark this answers is not in the document]
Comment (qshuai) [reply blockchain/Validation.go]: I do not find it in the current repository. Please point it out!

File: blockchain/Validation.go [hunk 11]
```diff
@@ -120,24 +169,29 @@ func IsCashHFEnabled(params *msg.BitcoinParams, medianTimePast int64) bool {
...
 
 	return true
 }
 
-func ContextualCheckBlock(params *msg.BitcoinParams, block *model.Block, state *model.ValidationState, pindexPrev *BlockIndex) bool {
+func ContextualCheckBlock(params *msg.BitcoinParams, block *model.Block, state *model.ValidationState,
+	pindexPrev *BlockIndex) bool {
+
 	nHeight := pindexPrev.Height + 1
 	if pindexPrev == nil {
```
[the remark this answers is not in the document]
Comment (qshuai) [reply blockchain/Validation.go]: get it

File: blockchain/Validation.go [hunk 12]
```diff
@@ -2248,19 +2325,476 @@ var (
...
 	tx *model.Tx, limitFree bool, missingInputs *bool, acceptTime int64, txReplaced *list.List,
-	overrideMempoolLimit bool, absurdFee btcutil.Amount, coinsToUncache []*model.OutPoint) bool {
+	overrideMempoolLimit bool, absurdFee btcutil.Amount, coinsToUncache []*model.OutPoint) (ret bool) {
+
+	//! notice missingInputs acts as a pointer to boolean type
+	// todo AssertLockHeld(cs_main)
+
+	ptx := tx
+	txid := ptx.TxHash()
+	if *missingInputs {
```
[the remark this answers is not in the document]
Comment (qshuai) [reply blockchain/Validation.go]: get it

File: blockchain/Validation.go [hunk 13]
```diff
@@ -2248,19 +2325,476 @@ var (
...
+	}
+
+	// Is it already in the memory pool?
+	if pool.Exists(txid) {
+		ret = state.Invalid(false, REJECT_ALREADY_KNOWN, "txn-already-in-mempool", "")
+		return
+	}
+
+	// Check for conflicts with in-memory transactions
+	defer func() {
```
[the remark this answers is not in the document]
Comment (qshuai) [reply blockchain/Validation.go]: defer func() is closure function， the lock will be immediately unlocked when this closure finishes.  It is important especially in a long function.

File: blockchain/Validation.go (hunk 1 above)
Comment (qshuai) [reply blockchain/Validation.go]: you are right

File: blockchain/Validation.go (hunk 2 above)
Comment (qshuai) [reply blockchain/Validation.go]: this logic convert maybe correct. let us discuss.

File: blockchain/Validation.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (qshuai) [reply blockchain/Validation.go]: I have correct it

File: blockchain/Validation.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (qshuai) [reply blockchain/Validation.go]: you are right!!!

File: blockchain/Validation.go (hunk 5 above)
Comment (qshuai) [reply blockchain/Validation.go]: you are right

**Comment by ludete** [comment]
OK

**Pull request merged**